python/ngen_cal/src/ngen/cal/agent.py

```diff
@@ -0,0 +1,117 @@
+            return 0
+    
+    @property
+    @abstractmethod
+    def model(self) -> 'Model':
+        pass
+
+    @property
+    @abstractmethod
+    def job(self) -> 'JobMeta':
+        pass
+    
+    def update_config(self, i: int, params: 'DataFrame', id: str):
+        """
+            For a given calibration iteration, i, update the input files/configuration to prepare for that iterations
+            calibration run.
+
+            parameters
+            ---------
+            i: int
+                current iteration of calibration
+            params: pandas.DataFrame
+                DataFrame containing the parameter name in `param` and value in `i` columns
+        """
+        return self.model.update_config(i, params, id, path=self.job.workdir)
+    
+    @property
+    def best_params(self) -> str:
+        return self.model.best_params
+
+class Agent(BaseAgent):
+
+    def __init__(self, model_conf, workdir: 'Path', log: bool=False, restart: bool=False, parameters: 'Optional[Mapping[str, Any]]' = {}):
+        self._workdir = workdir
+        self._job = None
+        if restart:
+            # find prior ngen workdirs
+            # FIXME if a user starts with an independent calibration strategy
+            # then restarts with a uniform strategy, this will "work" but probably shouldn't.
+            # it works cause the independent writes a param df for the nexus that uniform also uses,
+            # so data "exists" and it doesn't know its not conistent...
+            # Conversely, if you start with uniform then try independent, it will start back at
+            # 0 correctly since not all basin params can be loaded.
+            # There are probably some similar issues with explicit and independent, since they have
+            # similar data semantics
+            workdirs = list(Path.glob(workdir, model_conf['type']+"_*_worker"))
+            if( len(workdirs) > 1) :
+                print("More than one existing {} workdir, cannot restart")
+            elif( len(workdirs) == 1):
+                self._job = JobMeta(model_conf['type'], workdir, workdirs[0], log=log)
+
+        if(self._job is None):
+            self._job = JobMeta(model_conf['type'], workdir, log=log)
+        model_conf['workdir'] = self.job.workdir
+        self._model = Model(model=model_conf)
+        self._model.model.resolve_paths()
+        self._params = parameters
+    
+    @property
+    def parameters(self) -> 'Mapping[str, Any]':
+        return self._params
+
+    @property
+    def workdir(self) -> 'Path':
+        return self._workdir
+
+    @property
+    def job(self) -> 'JobMeta':
+        return self._job
+    
+    @property
+    def model(self) -> 'Model':
+        return self._model.model
+
+    @property
+    def cmd(self) -> str:
+        """
+            Proxy method to build command from contained model binary and args
+        """
+        return "{} {}".format(self.model.get_binary(), self.model.get_args())
```

Comment:
I'm not sure if this is the place, but it might be necessary to path resolve the binary location here... since the workers now run in a subdir, my config as follows doesn't find the binary:
```
    binary: "../ngen-serial"
```
This despite `workdir` being explicitly set to `./`.... which... I'm now not sure what that or the default value means anymore.

Reply:
Generally it seems to make relative paths in all the config files be impractical... that's annoying... but may not be easy to fix...
```
NGen Framework 0.1.0
Building Nexus collection
Building Catchment collection
libc++abi: terminating with uncaught exception of type std::runtime_error: Error opening forcing data dir './Input/forcing/' after 1 attempts: No such file or directory.
```

Reply:
Not real sure what the suggested change is...do you think resolving the relative binary path input will be sufficient?